### PR TITLE
Forbid calling findRenderObject on an unmounted element

### PR DESCRIPTION
**Pull request by dnfield**
Calling `context.findRenderObject()` on an unmounted state/element is always a bad idea.

We check this for `context.size`, but not `context.findRenderObject`. I noticed this because in my efforts to make render objects disposable, we actually null out the RO after unmounting, which made an internal test fail because it started getting a null value in a callback that used to not get a null value.

Updates docs, adds a test.

### CODE REVIEW

**Comment by dnfield** [comment]
This should not be landed until cl/377398477 has been submitted internally.

File: packages/flutter/test/widgets/framework_test.dart [hunk 1]
```diff
@@ -1619,6 +1619,32 @@ void main() {
...
+    expect(element.widget, isA<_StatefulLeaf>());
+    // Replace the widget tree to unmount the element.
+    await tester.pumpWidget(Container());
+
+    try {
+      element.findRenderObject();
+      fail('Expected exception');
+    } catch (e) {
+      // Make sure we only throw this type of error and no other.
+      final FlutterError error = e as FlutterError;
```
Comment (Hixie) [inline packages/flutter/test/widgets/framework_test.dart:1636]:
```suggestion
    } on FlutterError catch (e) {
```

File: packages/flutter/test/widgets/framework_test.dart [hunk 2]
```diff
@@ -1619,6 +1619,32 @@ void main() {
...
+        error.message,
+        equalsIgnoringHashCodes('''
+Cannot get renderObject of inactive element.
+In order for an element to have a valid renderObject, it must be active, which means it is part of the tree.
+Instead, this element is in the _ElementLifecycle.defunct state.
+The findRenderObject() method was called for the following element:
+  StatefulElement#00000(DEFUNCT)'''),
+      );
+    }
+  }, skip: kIsWeb);
```
Comment (Hixie) [inline packages/flutter/test/widgets/framework_test.dart:1647]: did you mean to skip this on web?

File: packages/flutter/test/widgets/framework_test.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/test/widgets/framework_test.dart:1636]: Ahh I had it this way and then mistakenly worried we'd somehow miss some other type of exception.

File: packages/flutter/test/widgets/framework_test.dart (hunk 2 above)
Comment (dnfield) [reply packages/flutter/test/widgets/framework_test.dart:1647]: No, copy paste error.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 3]
```diff
@@ -2455,6 +2455,11 @@ class InspectorSelection {
 
   Element? _currentElement;
   set currentElement(Element? element) {
+    if (element?.debugIsDefunct == true) {
+      _currentElement = null;
+      _current = null;
+      return;
+    }
```
Comment (dnfield) [inline packages/flutter/lib/src/widgets/widget_inspector.dart:2462]: /cc @jacob314

File: packages/flutter/test/widgets/framework_test.dart [hunk 4]
```diff
@@ -1619,6 +1619,30 @@ void main() {
...
+      expect(
+        e.message,
+        equalsIgnoringHashCodes('''
+Cannot get renderObject of inactive element.
+In order for an element to have a valid renderObject, it must be active, which means it is part of the tree.
+Instead, this element is in the _ElementLifecycle.defunct state.
+The findRenderObject() method was called for the following element:
+  StatefulElement#00000(DEFUNCT)'''),
+      );
+    }
```
Comment (goderbauer) [inline packages/flutter/test/widgets/framework_test.dart:1644]: Less error-prone pattern:

```dart
  expect(
    () => element.findRenderObject(),
    throwsA(isA<FlutterError>().having(
      (FlutterError error) => error.message,
      'message',
      'Cannot ger renderObject ...'
    ),
  );
```

File: packages/flutter/lib/src/widgets/framework.dart [hunk 5]
```diff
@@ -3864,7 +3865,23 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
...
+    assert(() {
+      if (_lifecycleState != _ElementLifecycle.active) {
+        throw FlutterError.fromParts(<DiagnosticsNode>[
+          ErrorSummary('Cannot get renderObject of inactive element.'),
+          ErrorDescription(
+            'In order for an element to have a valid renderObject, it must be '
+            'active, which means it is part of the tree.\n'
+            'Instead, this element is in the $_lifecycleState state.',
+          ),
+          describeElement('The findRenderObject() method was called for the following element'),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart:3878]: I wonder if we should add some kind of a hint around "If you're calling this from a State object, consider guarding it with State.mounted." to the error to make this a little more actionable?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/test/widgets/framework_test.dart (hunk 4 above)
Comment (dnfield) [reply packages/flutter/test/widgets/framework_test.dart:1644]: Done

File: packages/flutter/lib/src/widgets/framework.dart (hunk 5 above)
Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart:3880]: Done

**Comment by dnfield** [comment]
This will still have to wait on the internal CL, which will likely not get landed until Monday at the earliest due to an OOO reviewer.

File: packages/flutter/lib/src/widgets/widget_inspector.dart (hunk 3 above)
Comment (jacob314) [reply packages/flutter/lib/src/widgets/widget_inspector.dart:2462]: lgtm

**Pull request merged**